tests/system/src/test/java/test/util/ScreenshotCapture.java

```diff
@@ -0,0 +1,156 @@
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.util.Base64;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.imageio.ImageIO;
+import javafx.application.Platform;
+import javafx.embed.swing.SwingFXUtils;
+import javafx.geometry.Rectangle2D;
+import javafx.scene.image.WritableImage;
+import javafx.scene.robot.Robot;
+import javafx.stage.Screen;
+
+/**
+ * A utility to capture a screenshot via JavaFX {@link Robot}.
+ * <p>
+ * Example:
+ * <pre>
+ * // write a base-64 encoded screenshot to stderr
+ * ScreenshotCapture.writeScreenshot(System.err);
+ * </pre>
+ */
+public class ScreenshotCapture {
+    /**
+     * Captures a screenshot using JavaFX {@link Robot} in the PNG format.
+     * <p>
+     * This method can be called from any thread.  If called from a thread other than
+     * the JavaFX Application Thread, the current thread will be paused until the screenshot is taken.
+     *
+     * @return the byte array containing the screenshot
+     * @throws IOException when an I/O error occurs
+     */
+    public static byte[] takeScreenshot() throws IOException {
+        if (Platform.isFxApplicationThread()) {
+            return screenshotFX();
+        } else {
+            AtomicReference<Object> ref = new AtomicReference<>();
+            Util.runAndWait(() -> {
+                try {
+                    Object s = screenshotFX();
+                    ref.set(s);
+                } catch (IOException e) {
+                    ref.set(e);
+                }
+            });
+            Object result = ref.get();
+            if (result instanceof IOException e) {
+                throw e;
+            }
+            return (byte[])result;
+        }
+    }
+
+    /**
+     * Captures a screenshot using JavaFX {@link Robot} in the PNG format,
+     * then writes it in a Base-64 encoding to the specified {@code PrintStream}.
+     * <p>
+     * Example:
+     * <pre>
+     * Screenshot:
+     * data:image/png;base64,iVBORw0KGgoA...</pre>
+     *
+     * @param out the output print stream
+     */
+    public static void writeScreenshot(PrintStream out) {
+        out.println(ScreenshotCapture.takeScreenshotBase64("Screenshot:\ndata:image/png;base64,", null));
+    }
+
+    /**
+     * Captures a screenshot using JavaFX {@link Robot} in the PNG format,
+     * in the form of a Base-64 encoded {@code String}.
+     * <p>
+     * This method can be called from any thread.  If called from a thread other than
+     * the JavaFX Application Thread, the current thread will be paused until the screenshot is taken.
+     *
+     * @param prefix the string to append before the base-64 representation, or null
+     * @param postfix the string to append after the base-64 representation, or null
+     * @return the screenshot in Base-64 encoded PNG, or an error message
+     */
+    public static String takeScreenshotBase64(String prefix, String postfix) {
```

Comment:
Passing in the `data:image/png;base64,` as a prefix is odd. This method takes a screen shot and turns it into a base64-encoded PNG. Don't make the caller pass in something that has to match this internal behavior of this method. I would recommend removing the prefix and postfix entirely, and having this method return a simple base64-encoded PNG with a ``data:image/png;base64,` data URL. The caller can add any other String before or after this.

Reply:
I wanted to avoid creating multiple multi-megabyte strings.

The method to use is writeScreenshot(), which should be simple enough to use, it's the method mentioned at the class level javadoc.

Related: I was debating whether we should have two specific methods
```
writeScreenshotStdout();
writeSceeenshotStderr();
```
instead of passing `System:out` or `System::err`.  What do you think?

Reply:
My suggestion doesn't change anything about how many or what size strings are allocated in the common case. All it does is moves which method tacks on the needed "data:image/png;base64," piece from the caller to the method that actually creates the PNG.

As for two methods vs one, I like the flexibility as you have it, although really only the System.err version is likely to be used (with gradle as the test runner, writing it to stdout isn't generally as useful).